### PR TITLE
effectful-typed-process

### PR DESCRIPTION
**Pull request by mmhat**
This PR adds the `effectful-typed-process` package as announced in https://github.com/arybczak/effectful/pull/25#issuecomment-887881309 .

### CODE REVIEW

**Comment by mmhat** [comment]
@Kleidukos I think this one can be merged. You might want to have a quick look here if everything is well and merge it if that's the case.

**Comment by mmhat** [comment]
As a side note: I kept some commented functions in `Effectful.Process.Typed`. Those are the ones using `STM` and the STM effect got removed in https://github.com/arybczak/effectful/commit/8057cdcf5f96170db16d14e167f259b9e7026bbd. I cannot recall if that removal is just temporary or if it will stay this way - Maybe @arybczak can help clarifying the situation. Depending on the future course of the effectful library we could either remove or uncomment those.

File: effectful-typed-process/src/Effectful/Process/Typed.hs [hunk 1]
```diff
@@ -0,0 +1,221 @@
+{-# OPTIONS_GHC -fno-warn-dodgy-imports #-}
+
+module Effectful.Process.Typed
```
Comment (Kleidukos) [inline effectful-typed-process/src/Effectful/Process/Typed.hs:3]: This got me thinking, because in the other libraries, we use the pattern <Original.Namespace>.Effect.

But maybe we could indeed organise all this under the Effectful.* namespace. Do you have any opinion, @mmhat ?

Comment (mmhat) [reply effectful-typed-process/src/Effectful/Process/Typed.hs:3]: @Kleidukos Yes, I noticed that and I was quite unsure which way to go here. At the end I went with the `Effectful.*` namespace to maintain consistency with the modules in the `effectful`/`effectful-core` packages. As a side effect the import lists look pretty nice since tooling will group all the effectful imports together.

**TLDR**; I'd go with the Effectful namespace, but I do not have a really strong opinion regarding that :slightly_smiling_face:

Comment (Kleidukos) [reply effectful-typed-process/src/Effectful/Process/Typed.hs:3]: Yep, let's go with it.

**Comment by Kleidukos** [comment]
@mmhat Do you need help with the test suite?

**Comment by mmhat** [comment]
> Do you need help with the test suite?

Oh, thank you very much! I added a first draft of the test suite and feedback is very welcome. In particular, if you have any idea how to write portable tests for a library like that I'd highly appreciate those.

File: effectful-typed-process/src/Effectful/Process/Typed.hs (hunk 1 above)
Comment (mmhat) [reply effectful-typed-process/src/Effectful/Process/Typed.hs:3]: @Kleidukos I could prepare some PRs to adjust the namespacing of the other libraries if you are ok with that.

Comment (Kleidukos) [reply effectful-typed-process/src/Effectful/Process/Typed.hs:3]: I am absolutely OK with it :)

File: effectful-typed-process/effectful-typed-process.cabal [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  exposed-modules:
+    Effectful.Process.Typed
+
+  build-depends:    base <= 4.17
+                  , bytestring
+                  , typed-process
+                  , effectful-core
+
+test-suite effectful-cache-test
```
Comment (Kleidukos) [inline effectful-typed-process/effectful-typed-process.cabal:45]:
```suggestion
test-suite effectful-process-test
```

**Review by Kleidukos — CHANGES_REQUESTED** [review]

File: effectful-typed-process/effectful-typed-process.cabal (hunk 2 above)
Comment (mmhat) [reply effectful-typed-process/effectful-typed-process.cabal]: Uh, thank you! :smile: 
I fixed that.

File: effectful-typed-process/test/Main.hs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      it "readProcessStderr_" $ do
+        let pc = shell "printf 'Output' >&2"
+            action = runEff . runTypedProcess $ do
+              readProcessStderr_ pc
+        action `shouldReturn` "Output"
+
+    describe "Exit codes" $ do
+
+      it "runProcess_" $ do
+        let pc = proc "/bin/env" ["false"]
```
[the remark this answers is not in the document]
Comment (mmhat) [reply effectful-typed-process/test/Main.hs]: Thank you, I changed that.

**Comment by mmhat** [comment]
@Kleidukos I guess if there is nothing else we can merge this?

**Review by Kleidukos — APPROVED** [review]
LGTM! Thank you very much for this invaluable contribution. :)
